pkg/metric_storage/metric_storage.go

```diff
@@ -369,14 +362,17 @@ func (m *MetricStorage) SendBatch(ops []operation.MetricOperation, labels map[st
 func (m *MetricStorage) ApplyOperation(op operation.MetricOperation, commonLabels map[string]string) {
 	labels := MergeLabels(op.Labels, commonLabels)
 
-	if op.Add != nil {
-		m.CounterAdd(op.Name, *op.Add, labels)
+	if op.Action == "add" && op.Value != nil {
+		m.CounterAdd(op.Name, *op.Value, labels)
 		return
 	}
-	if op.Set != nil {
-		m.GaugeSet(op.Name, *op.Set, labels)
+	if op.Action == "set" && op.Value != nil {
+		m.GaugeSet(op.Name, *op.Value, labels)
```

Comment:
It is better to check both conditions.

```suggestion
	if op.Set != nil {
		m.GaugeSet(op.Name, *op.Set, labels)
		return
	}
	if op.Action == "set" && op.Value != nil {
		m.GaugeSet(op.Name, *op.Value, labels)
```